crates/sui-types/src/crypto.rs

```diff
@@ -203,53 +183,77 @@ pub fn random_key_pairs(num: usize) -> Vec<KeyPair> {
 
 // TODO: get_key_pair() and get_key_pair_from_bytes() should return KeyPair only.
 // TODO: rename to random_key_pair
-pub fn get_key_pair() -> (SuiAddress, KeyPair) {
+pub fn get_key_pair<KP: KeypairTraits>() -> (SuiAddress, KP) {
     get_key_pair_from_rng(&mut OsRng)
 }
 
 /// Generate a keypair from the specified RNG (useful for testing with seedable rngs).
-pub fn get_key_pair_from_rng<R>(csprng: &mut R) -> (SuiAddress, KeyPair)
+pub fn get_key_pair_from_rng<KP: KeypairTraits, R>(csprng: &mut R) -> (SuiAddress, KP)
 where
     R: rand::CryptoRng + rand::RngCore,
 {
-    let kp = KeyPair::generate(csprng);
+    let kp = KP::generate(csprng);
     (kp.public().into(), kp)
 }
 
 // TODO: C-GETTER
-pub fn get_key_pair_from_bytes(bytes: &[u8]) -> SuiResult<(SuiAddress, KeyPair)> {
-    let priv_length = <KeyPair as narwhal_crypto::traits::KeyPair>::PrivKey::LENGTH;
-    let sk =
-        PrivateKey::from_bytes(&bytes[..priv_length]).map_err(|_| SuiError::InvalidPrivateKey)?;
-    let kp: KeyPair = sk.into();
+pub fn get_key_pair_from_bytes<KP: KeypairTraits>(bytes: &[u8]) -> SuiResult<(SuiAddress, KP)> {
+    let priv_length = <KP as KeypairTraits>::PrivKey::LENGTH;
+    let sk = <KP as KeypairTraits>::PrivKey::from_bytes(&bytes[..priv_length])
+        .map_err(|_| SuiError::InvalidPrivateKey)?;
+    let kp: KP = sk.into();
     if kp.public().as_ref() != &bytes[priv_length..] {
         return Err(SuiError::InvalidAddress);
     }
     Ok((kp.public().into(), kp))
 }
 
-// TODO: replace this with a byte interpretation based on multicodec
-pub const SUI_SIGNATURE_LENGTH: usize = PublicKey::LENGTH + AccountSignature::LENGTH;
+//
+// Account Signatures
+//
 
-#[serde_as]
-#[derive(Eq, PartialEq, Copy, Clone, Serialize, Deserialize, JsonSchema)]
-pub struct Signature(
-    #[schemars(with = "Base64")]
-    #[serde_as(as = "Readable<Base64, Bytes>")]
-    [u8; SUI_SIGNATURE_LENGTH],
-);
+// 1. Eddsa
+// 2. Ecdsa
+// Enums for Signatures
+const FLAG_LENGTH: usize = 1;
+
+#[enum_dispatch]
+#[derive(Clone, Serialize, Deserialize, JsonSchema)]
+pub enum Signature {
+    Secp256k1SuiSignature,
+}
+
+// Can refactor this with a library
+impl Signature {
+    pub fn new<T>(value: &T, secret: &dyn signature::Signer<Signature>) -> Signature
+    where
+        T: Signable<Vec<u8>>,
+    {
+        let mut message = Vec::new();
+        value.write(&mut message);
+        secret.sign(&message)
+    }
+}
 
 impl AsRef<[u8]> for Signature {
     fn as_ref(&self) -> &[u8] {
-        self.0.as_ref()
+        match self {
+            Signature::Secp256k1SuiSignature(sig) => sig.as_ref(),
+        }
     }
 }
 
 impl signature::Signature for Signature {
```

Comment:
should we rename `enum Signature` to `SignatureType` or `SigningType`? its quite an overloaded name

Reply:
Perhaps we can change it to AccountSignature, after the current 'AccountSignature' is refactored away?